services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/api/containers.py

```diff
@@ -0,0 +1,241 @@
+    background_tasks: BackgroundTasks,
+    settings: DynamicSidecarSettings = Depends(get_settings),
+    shared_store: SharedStore = Depends(get_shared_store),
+) -> Union[List[str], Dict[str, Any]]:
+    """Expects the docker-compose spec as raw-body utf-8 encoded text"""
+
+    # stores the compose spec after validation
+    body_as_text = (await request.body()).decode("utf-8")
+
+    try:
+        shared_store.compose_spec = await validate_compose_spec(
+            settings=settings, compose_file_content=body_as_text
+        )
+        shared_store.container_names = assemble_container_names(
+            shared_store.compose_spec
+        )
+    except InvalidComposeSpec as e:
+        logger.warning("Error detected %s", traceback.format_exc())
+        raise HTTPException(status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e)) from e
+
+    # run docker-compose in a background queue and return early
+    background_tasks.add_task(task_docker_compose_up, settings, shared_store)
+
+    return shared_store.container_names
+
+
+@containers_router.post("/containers:down", response_class=PlainTextResponse)
+async def runs_docker_compose_down(
+    command_timeout: float = Query(
+        10.0, description="docker-compose down command timeout default"
+    ),
+    settings: DynamicSidecarSettings = Depends(get_settings),
+    shared_store: SharedStore = Depends(get_shared_store),
+) -> Union[str, Dict[str, Any]]:
+    """Removes the previously started service
+    and returns the docker-compose output"""
+
+    stored_compose_content = shared_store.compose_spec
+    if stored_compose_content is None:
+        raise HTTPException(
+            status.HTTP_422_UNPROCESSABLE_ENTITY,
+            detail="No spec for docker-compose down was found",
+        )
+
+    finished_without_errors, stdout = await remove_the_compose_spec(
+        shared_store=shared_store,
+        settings=settings,
+        command_timeout=command_timeout,
+    )
+
+    if not finished_without_errors:
+        raise HTTPException(status.HTTP_500_INTERNAL_SERVER_ERROR, detail=stdout)
+
+    return stdout
+
+
+@containers_router.get(
+    "/containers",
+    responses={
+        status.HTTP_500_INTERNAL_SERVER_ERROR: {"description": "Errors in container"}
+    },
+)
+async def containers_docker_inspect(
+    only_status: bool = Query(
+        False, description="if True only show the status of the container"
+    ),
+    shared_store: SharedStore = Depends(get_shared_store),
+) -> Dict[str, Any]:
+    """
+    Returns entire docker inspect data, if only_state is True,
+    the status of the containers is returned
+    """
+
+    def _format_result(container_inspect: Dict[str, Any]) -> Dict[str, Any]:
+        if only_status:
+            container_state = container_inspect.get("State", {})
+
+            # pending is another fake state use to share more information with the frontend
+            return {
+                "Status": container_state.get("Status", "pending"),
```

Comment:
are you sure the docker engine does not return anything special when pulling/starting?

Reply:
Yes, I've checked when putting this comment. It's needed for the frontend because we are mapping from task to container statues which do not overlap completely